### PR TITLE
address review comments

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2936]: make a method for it, so that we can reuse it in `SparkSqlAstBuilder`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: `visitReplaceTableHeader` simply return 3 `false` for these 3 properties. So it's simpler to use assert here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3106]: Previously, when table columns list is not specified, we ignore the partition columns with data type. It was fine before syntax merging, as there was no partition columns with data type in REPLACE TABLE. But now it's better to make it consistent with CREATE TABLE. I also added test to check it: https://github.com/rdblue/spark/pull/8/files#diff-b9e91f767e5562861565b0ce78759af3bcb7fff405a81e928894641147db2ae4R293

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala:66]: it's a bit weird to print scala Option directly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala:88]: otherwise the assert below is useless.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:279]: I did some refactoring, to avoid duplicated code between CREATE TABLE and CTAS, `buildCatalogTable` and `buildHiveCatalogTable`

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:647]: This method closely follow the original logic of creating hive table in `SparkSqlAstBuilder`

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:667]: The logic here is roughly merging 3 serde infos: the one user-specified, the one inferred from STORED AS, and the default one.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:366]: We can get rid of duplicated code here once we move CREATE TABLE LIKE and INSERT DIRECTORY to v2 command.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:594]: I don't know why INSERT DIRECTORY considers the default serde info but CREATE TABLE LIKE does not. I'm going to fix it later and keep the behavior unchanged here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: The other error message is more clear and direct. "Create partitioned table as select" is really a case of "create table as select" with partitions and should not be a distinct noun as it is used here. Also, it isn't clear what the "target table" is. There is no target because there is only one table and that one is being created.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I think it is bad practice for a method to assume it will get certain results from other methods. While using an assert handles correctness, if there is a change that violates the assertion, a user would get a nearly unusable error.

I think this change should be reverted so that the error messages are meaningful.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: The other error message is more clear and should be used instead of this one to be consistent with create.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala:125]: It would be better to include just the keys that conflicted.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Changing to infix notation isn't needed and is not recommended: https://docs.scala-lang.org/style/method-invocation.html#arity-1-infix-notation

Let's revert that change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Looks like this should use `_._2` to use the output format rather than the input format twice. This is one of the reasons to use a matcher instead of the `_1` way to access the fields, but it isn't a big problem.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:391]: This is okay, but my intent was to avoid mixing parsing logic and validation where possible. The parser should return what was parsed to Spark, which should decide whether it is supported.

I don't think this is a blocker because it is in the SparkSqlParser instead of the one in catalyst. We can fix this when moving these commands to v2.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala]: This should throw a helpful exception with a real message. Before it was `throw new AnalysisException(s"Cannot create table with both USING $provider and ${serde.describe}")`

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala]: No need for the operator since this isn't a key/value pair. It is more clear to use `(nonHiveStorageFormat, provider.get)`.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:279]: Looks fine to me. I like that it should have fewer changes from master.

**Review by rdblue — CHANGES_REQUESTED** [review]
@cloud-fan, it mostly looks good but I did find a few things to change and at least one bug from using `_._1` instead of `_._2`. Let me know when you've had time to update this.

**Comment by cloud-fan** [comment]
@rdblue thanks for the suggestions! Pushed a commit to update it, please take another look, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I think this should use the original error message, "CREATE OR REPLACE TEMPORARY TABLE ..., use CREATE TEMPORARY VIEW instead".

Using `operationNotAllowed` means that the message is prefixed with "Operation not allowed: ", so adding "is not supported yet." is not helpful and just makes the message harder to read. In addition, "yet" implies that this will be supported and there is no reason to do so.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: This should use the other error message, "CREATE TEMPORARY TABLE ... AS ..., use CREATE TEMPORARY VIEW instead".

As I noted on the similar case, "is not supported yet" is both redundant and misleading. I don't think that Spark intends to implement `CREATE TEMPORARY TABLE`. Even if it may be implemented, it has not been supported for years, so there is no value in implying that it will be supported.

Please update to use the simpler and clearer error mesage.

**Review by rdblue — CHANGES_REQUESTED** [review]
I think the only things that need to be fixed are the error messages that were changed and are now longer and less clear.

**Comment by rdblue** [comment]
Thanks, @cloud-fan. I've merged this.

**Pull request merged**
